saq/queue.py

```diff
@@ -403,3 +439,127 @@ async def enqueue(self, job_or_func, **kwargs):
 
         logger.info("Enqueuing %s", job)
         return job
+
+    async def apply(self, job_or_func, timeout=None, **kwargs):
+        """
+        Enqueue a job and wait for its result.
+
+        If the job is successful, this returns its result.
+        If the job is unsuccessful, this raises a JobError.
+
+        Example::
+            try:
+                assert await queue.apply("add", a=1, b=2) == 3
+            except JobError:
+                print("job failed")
+
+        job_or_func: Same as Queue.enqueue
+        kwargs: Same as Queue.enqueue
+        """
+        results = await self.map(job_or_func, timeout=timeout, iter_kwargs=[kwargs])
+        return results[0]
+
+    async def map(self, job_or_func, iter_kwargs, timeout=None, **kwargs):
+        """
+        Enqueue multiple jobs and collect all of their results.
+
+        This raises a JobError if any of the jobs are unsuccessful.
+
+        This waits for job results in the order they are enqueued, so JobExceptions are
+        not necessarily raised immediately.
+
+        Example::
+            try:
+                assert await queue.map(
+                    "add",
+                    [
+                        {"a": 1, "b": 2},
+                        {"a": 3, "b": 4},
+                    ]
+                ) == [3, 7]
+            except JobError:
+                print("any of the jobs failed")
+
+        job_or_func: Same as Queue.enqueue
+        iter_kwargs: Enqueue a job for each item in this sequence. Each item is the same
+            as kwargs for Queue.enqueue.
+        timeouts: Total seconds to wait for all jobs to complete. If None or 0, wait forever.
+        kwargs: Default kwargs for all jobs. These will be overridden by those in iter_kwargs.
+        """
+        jobs = [self.prepare_job(job_or_func, **{"timeout": timeout, **kwargs, **kw}) for kw in iter_kwargs]
+        jobs_by_id = {job.id: job for job in jobs}
+
+        done = 0
+        failed_job = None
+
+        async def callback(_id, status):
+            nonlocal done, failed_job
```

Comment:
not super into this section

maybe it's more robust to store jobs as a set and just remove them when we receive them here as opposed to relying on a counter

for the failed job, would it be simplified if we just don't care if jobs fail? that's on the caller

Reply:
I'll switch it to a set of job ids.

Do you mean we don't care if jobs fail within this callback? Or we don't care if jobs fail within the `map` method?

Reply:
both?

Reply:
but not a fan of having to do nonlocal failed_job and also, you only track when one job is failed and not others, seems just weird, could it be cleaner to ignore failures in the map method?

Reply:
`map` is essentially the same thing as [asyncio.gather](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather).

Like `gather` returning the awaitable return values and the the tasks, `map` returns the job results and not the jobs themselves.

`gather` makes the behavior we're discussing configurable via the `return_exceptions` argument:

> If return_exceptions is False (default), the first raised exception is immediately propagated to the task that awaits on gather(). Other awaitables in the aws sequence won’t be cancelled and will continue to run.
>
> If return_exceptions is True, exceptions are treated the same as successful results, and aggregated in the result list.

Maybe we should copy asyncio and add a `return_errors` parameter?